### PR TITLE
support pir_p_to_s_reshard and test

### PR DESCRIPTION
**Pull request by smile2game**
### PR Category
Auto Parallel

### PR Types
New features

### Description
本次 PR 添加了 `p_to_r_unbalanced_reshard` 函数，用于处理不平衡的重新分片操作。该函数旨在优化数据分布，确保在资源分配不均的情况下，能够高效地进行数据重新分片。主要改进为：

- 实现了基于负载的动态分片算法，提高了系统的扩展性和稳定性。
- 减少了数据迁移过程中对系统性能的影响，提升了整体运行效率。
- 增强了错误处理机制，确保在重新分片过程中能够及时捕捉并处理异常情况。

通过引入 `p_to_r_unbalanced_reshard` 函数，系统在面对不平衡负载时能够更灵活地进行资源调整，提升了整体性能和可靠性。

### CODE REVIEW

File: test/auto_parallel/pir/pir_reshard_p_to_s.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+                else:
+                    assert op_result_dist_attr.dims_mapping == [-1, 0]
+                assert op_result_dist_attr.partial_status == {}
+
+                op_value = op.result(0)
+                assert op_value.is_dense_tensor_type()
+                assert op_value.is_dist_dense_tensor_type()
+                assert op_value.dist_attr().process_mesh == self._mesh
+                if self._shard == 0:
+                    pass
```
Comment (jeff41404) [inline test/auto_parallel/pir/pir_reshard_p_to_s.py]: we should check `op_value.dist_attr().dims_mapping == [0, -1]`?

File: test/auto_parallel/pir/pir_reshard_p_to_s.py [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+                else:
+                    assert op_result_dist_attr.dims_mapping == [-1, 0]
```
Comment (jeff41404) [inline test/auto_parallel/pir/pir_reshard_p_to_s.py]: we should check `op_value.dist_attr().dims_mapping == [-1, 0]`? not `op_result_dist_attr.dims_mapping`

File: test/auto_parallel/pir/pir_reshard_p_to_s.py [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+                assert op_result_dist_attr.partial_status == {}
+
+                op_value = op.result(0)
+                assert op_value.is_dense_tensor_type()
+                assert op_value.is_dist_dense_tensor_type()
+                assert op_value.dist_attr().process_mesh == self._mesh
+                if self._shard == 0:
+                    pass
```
Comment (jeff41404) [inline test/auto_parallel/pir/pir_reshard_p_to_s.py]: we should check `op_value.dist_attr().dims_mapping == [0, -1]`?

File: test/auto_parallel/pir/pir_reshard_p_to_s.py [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+                else:
+                    assert op_result_dist_attr.dims_mapping == [-1, 0]
```
Comment (jeff41404) [inline test/auto_parallel/pir/pir_reshard_p_to_s.py]: we should check `op_value.dist_attr().dims_mapping == [-1, 0]`? not `op_result_dist_attr.dims_mapping`

File: test/auto_parallel/pir/pir_reshard_p_to_s.py [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+                op_result_dist_attr = op.dist_attr.result(
+                    0
+                ).as_tensor_dist_attr()
+                assert op_result_dist_attr.process_mesh == self._mesh
+                assert op_result_dist_attr.dims_mapping == [-1, -1]
+                assert op_result_dist_attr.partial_status == {}
+                op_value = op.result(0)
+                assert op_value.is_dense_tensor_type()
+                assert op_value.is_dist_dense_tensor_type()
+                assert op_value.dist_attr().process_mesh == self._mesh
```
Comment (jeff41404) [inline test/auto_parallel/pir/pir_reshard_p_to_s.py:281]: we should also check `op_value.dist_attr().dims_mapping`

**Comment by jeff41404** [comment]
please add  unittest in [test/auto_parallel/pir/CMakeLists.txt](https://github.com/PaddlePaddle/Paddle/blob/develop/test/auto_parallel/pir/CMakeLists.txt) to solve issue of PR-CI-Coverage

File: python/paddle/distributed/auto_parallel/static/reshard_funcs/p_to_s_reshard_func.py [hunk 6]
```diff
@@ -44,58 +44,183 @@ def is_suitable(self, src_dist_attr, dst_dist_attr):
...
         assert (
             src_reduce_type == paddle.base.core.ReduceType.kRedSum
         ), f"The p to s reshard func only support sum op, but received {src_reduce_type}"
-
-        chunk_id = -1
-        if src_value.get_defining_op().dist_attr:
-            chunk_id = src_value.get_defining_op().dist_attr.chunk_id
-
         split_axis = dst_dist_attr.dims_mapping.index(0)
+        permulate = False
```
Comment (pkuzyc) [inline python/paddle/distributed/auto_parallel/static/reshard_funcs/p_to_s_reshard_func.py:52]: typo? permute

Comment (pkuzyc) [inline python/paddle/distributed/auto_parallel/static/reshard_funcs/p_to_s_reshard_func.py]: typo? permute

File: python/paddle/distributed/auto_parallel/static/reshard_funcs/p_to_s_reshard_func.py [hunk 7]
```diff
@@ -44,58 +44,183 @@ def is_suitable(self, src_dist_attr, dst_dist_attr):
...
+                tmp_dst_perm_type = paddle.base.libpaddle.pir.cvt_to_dist_type(
+                    dst_value.type(), dst_dist_attr
+                )
+                dst_value.set_type(tmp_dst_perm_type)
+
+                dst_value.get_defining_op().dist_attr = (
+                    paddle.base.libpaddle.pir.create_op_dist_attribute(
+                        dst_dist_attr.process_mesh,
+                        [dst_dist_attr],
+                        [dst_dist_attr],
```
Comment (pkuzyc) [inline python/paddle/distributed/auto_parallel/static/reshard_funcs/p_to_s_reshard_func.py]: The 'dist_attr' seems wrong. The 2nd and 3rd argument of 'create_op_dist_attribute' are the distributed attribute of the op's operands and results. Here operand's dist_attr is set with the result's dist_attr

File: test/auto_parallel/pir/test_pir_reshard_p_to_s.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
+# Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
```
Comment (pkuzyc) [inline test/auto_parallel/pir/test_pir_reshard_p_to_s.py]: Add this unit test file to CMakeList.txt. Fix the year 2023 -> 2024

**Review by jeff41404 — APPROVED** [review]
LGTM

**Review by pkuzyc — APPROVED** [review]
LGTM

**Pull request merged**